File: kotlin/tutorials/tour-of-orchestration-kotlin/src/main/kotlin/my/example/concurrenttasks/service.kt
import dev.restate.sdk.Context
import dev.restate.sdk.DurableFuture
import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.Service
import dev.restate.sdk.http.vertx.RestateHttpServer
import dev.restate.sdk.kotlin.endpoint
import java.util.*

data class SubscriptionRequest(
    val userId: String,
    val creditCard: String,
    val subscriptions: Array<String>
)

data class SubscriptionResult(
    val success: Boolean,
    val paymentRef: String
)

// Mock function to create recurring payment
suspend fun createRecurringPayment(creditCard: String, paymentId: String): String {
    return "payRef-${UUID.randomUUID()}"
}

// Mock function to create subscription
suspend fun createSubscription(userId: String, subscription: String, paymentRef: String) {
    println("Creating subscription for user: $userId, subscription: $subscription, paymentRef: $paymentRef")
}

@Service
class ParallelSubscriptionService {

    @Handler
    suspend fun add(ctx: Context, req: SubscriptionRequest): SubscriptionResult {
        val paymentId = ctx.random().nextUUID().toString()
        
        val payRef = ctx.runBlock(
            "pay",
            String::class
        ) {
            createRecurringPayment(req.creditCard, paymentId)
        }

        // Start all subscriptions in parallel
        val subscriptionFutures = mutableListOf<DurableFuture<Unit>>()
        for (subscription in req.subscriptions) {
            val future = ctx.runAsync("add-$subscription") {
                createSubscription(req.userId, subscription, payRef)
            }
            subscriptionFutures.add(future)
        }

        // Wait for all subscriptions to complete
        DurableFuture.all(subscriptionFutures).await()

        return SubscriptionResult(true, payRef)
    }
}

fun main() {
    RestateHttpServer.listen(endpoint {
        bind(ParallelSubscriptionService())
    })
}

File: python/tutorials/tour-of-orchestration-python/app/concurrenttasks/service.py
import restate
import asyncio

from app.utils import create_recurring_payment, create_subscription
from app.types import SubscriptionRequest, SubscriptionResult


parallel_subscription_service = restate.Service("ParallelSubscriptionService")


@parallel_subscription_service.handler()
async def add(ctx: restate.Context, req: SubscriptionRequest) -> SubscriptionResult:
    payment_id = str(ctx.uuid())
    pay_ref = await ctx.run_typed(
        "pay",
        create_recurring_payment,
        credit_card=req.credit_card,
        payment_id=payment_id,
    )

    # Start all subscriptions in parallel using asyncio.gather
    subscription_tasks = []
    for subscription in req.subscriptions:
        task = ctx.run_typed(
            f"add-{subscription}",
            create_subscription,
            user_id=req.user_id,
            subscription=subscription,
            payment_ref=pay_ref,
        )
        subscription_tasks.append(task)

    # Wait for all subscriptions to complete
    await asyncio.gather(*subscription_tasks)

    return SubscriptionResult(success=True, payment_ref=pay_ref)
